tool resilience: pare-test/build/npm misbehave when worktree has no node_modules

Three related issues observed in a single session today while working in a fresh `git worktree add ...` with no `node_modules`. All three tools should at minimum detect the missing-deps state and return a clear error; one of them (`pare-npm__install`) silently lies about success.

## 1. `pare-npm__install` returns `added: 0, changed: 0` but installs nothing

```jsonc
mcp__pare-npm__install({
  path: "/tmp/pare-worktree-fix-839",
  frozenLockfile: true,
  ignoreScripts: false,
  // ...
})
// → { added: 0, removed: 0, changed: 0, packageManager: "pnpm", lockfileChanged: false }
```

But `node_modules/` did not exist before *or* after the call. Running `pnpm install --frozen-lockfile` directly in the same directory installed everything in 2.5s (`Done in 2.5s using pnpm v10.29.2`).

Likely cause: the runner is parsing pnpm's "Already up to date" / "Lockfile is up to date" message and short-circuiting, but it should be actually invoking install — pnpm itself does install when `node_modules` is absent regardless of lockfile status.

**Severity: high** — a tool that returns success while doing nothing is much worse than a tool that errors. Subsequent `tsc`/`test` calls then fail with confusing errors (see #2 and #3 below) because the user thinks install succeeded.

## 2. `pare-test__run` returns `Expected property name or '}' in JSON at position 2` when deps are missing

```jsonc
mcp__pare-test__run({ path: "/tmp/pare-worktree-fix-839/packages/shared", ... })
// → Error: Expected property name or '}' in JSON at position 2
```

This is the same error noted in passing in #839's body. The fix: detect the "node_modules missing" / "vitest binary not found" case and return a typed error like `framework binary not found — run "pnpm install"` instead of letting the JSON parser blow up on whatever the failed runner emitted.

## 3. `pare-build__tsc` returns `success: false` with empty diagnostics

```jsonc
mcp__pare-build__tsc({ path: "/tmp/pare-worktree-fix-839/packages/shared", noEmit: true })
// → { success: false, diagnostics: [], errors: 0, warnings: 0 }
```

`success: false` with `errors: 0` and `diagnostics: []` is contradictory — the consumer can't tell what went wrong. Same root cause: tsc binary wasn't on PATH because `node_modules/.bin/tsc` didn't exist. After `pnpm install`, the same call returned `{ success: true, errors: 0, warnings: 0, diagnostics: [] }` correctly.

Should return a typed `tsc binary not found — run "pnpm install"` error.

## Common fix pattern

All three are instances of the same class: the underlying CLI failed in a way the parser/wrapper didn't anticipate, and the error surface degraded into something incoherent. A small `if (!existsSync(join(path, "node_modules", ".bin", BINARY))) throw new Error("...")` guard at the top of each runner would catch the most common case cleanly. Or — simpler — check for the well-known stderr signatures (`Command "tsc" not found`, `ERR_PNPM_RECURSIVE_EXEC_FIRST_FAIL`, etc.) and translate them.

## Repro

```bash
cd /tmp && git worktree add /tmp/pare-test-worktree -b throwaway main
# do NOT run pnpm install
# then call any of: pare-test__run, pare-build__tsc, pare-npm__install
```

All three observed in the session that produced PR #841 / closed PR #832.

--- a/.changeset/fix-missing-node-modules-typed-errors.md
+++ b/.changeset/fix-missing-node-modules-typed-errors.md
@@ -1,0 +1,11 @@
+---
+"@paretools/npm": patch
+"@paretools/test": patch
+"@paretools/build": patch
+---
+
+Detect missing `node_modules/` and surface typed errors instead of silently lying or returning incoherent diagnostics. See #842.
+
+- `npm.install`: after the package manager exits, verify a `node_modules/` directory was actually created (skipped for `dryRun` and `global`). If not, throw a typed error with `pm`, `cwd`, `exitCode`, and the last 5 stderr lines instead of returning `{ added: 0, removed: 0, changed: 0 }` as success.
+- `test.run`: before invoking the test framework, verify the JS framework binary (`jest` / `vitest` / `mocha`) is resolvable from a parent `node_modules/.bin/`. If missing, throw a typed `<framework> binary not found at <path> — try running "pnpm install"` error instead of crashing later with `Expected property name or '}' in JSON at position 2`. (pytest is unaffected — it's invoked via `python -m pytest`.)
+- `build.tsc`: same guard for `tsc` — return a typed not-found error instead of contradictory `{ success: false, errors: 0, diagnostics: [] }` output.

--- a/packages/server-build/__tests__/binary-check.test.ts
+++ b/packages/server-build/__tests__/binary-check.test.ts
@@ -1,0 +1,49 @@
+import { describe, it, expect, beforeEach, afterEach } from "vitest";
+import { mkdtempSync, rmSync, mkdirSync, writeFileSync, chmodSync } from "node:fs";
+import { tmpdir } from "node:os";
+import { join } from "node:path";
+import { assertBinaryAvailable } from "../src/lib/build-runner.js";
+
+describe("assertBinaryAvailable (build)", () => {
+  let tmp: string;
+
+  beforeEach(() => {
+    tmp = mkdtempSync(join(tmpdir(), "pare-build-binarycheck-"));
+  });
+
+  afterEach(() => {
+    rmSync(tmp, { recursive: true, force: true });
+  });
+
+  it("throws a typed error when node_modules/.bin/<binary> is missing", () => {
+    expect(() => assertBinaryAvailable(tmp, "tsc")).toThrowError(
+      /tsc binary not found at .*node_modules.*\.bin.*tsc.*— try running "pnpm install"/,
+    );
+  });
+
+  it("includes the cwd in the error so the consumer can locate the workspace", () => {
+    expect(() => assertBinaryAvailable(tmp, "tsc")).toThrowError(new RegExp(tmp));
+  });
+
+  it("succeeds when node_modules/.bin/<binary> exists in cwd", () => {
+    const binDir = join(tmp, "node_modules", ".bin");
+    mkdirSync(binDir, { recursive: true });
+    const binPath = join(binDir, "tsc");
+    writeFileSync(binPath, "#!/usr/bin/env node\n");
+    chmodSync(binPath, 0o755);
+    expect(() => assertBinaryAvailable(tmp, "tsc")).not.toThrow();
+  });
+
+  it("succeeds when the binary lives in an ancestor's node_modules (workspace hoisting)", () => {
+    const binDir = join(tmp, "node_modules", ".bin");
+    mkdirSync(binDir, { recursive: true });
+    const binPath = join(binDir, "tsc");
+    writeFileSync(binPath, "#!/usr/bin/env node\n");
+    chmodSync(binPath, 0o755);
+
+    const nested = join(tmp, "packages", "child");
+    mkdirSync(nested, { recursive: true });
+
+    expect(() => assertBinaryAvailable(nested, "tsc")).not.toThrow();
+  });
+});

--- a/packages/server-build/__tests__/binary-check.test.ts
+++ b/packages/server-build/__tests__/binary-check.test.ts
@@ -22,7 +22,9 @@ describe("assertBinaryAvailable (build)", () => {
   });
 
   it("includes the cwd in the error so the consumer can locate the workspace", () => {
-    expect(() => assertBinaryAvailable(tmp, "tsc")).toThrowError(new RegExp(tmp));
+    // Use string (substring match) rather than RegExp — Windows tmp paths
+    // contain backslashes that would be interpreted as regex escapes.
+    expect(() => assertBinaryAvailable(tmp, "tsc")).toThrowError(tmp);
   });
 
   it("succeeds when node_modules/.bin/<binary> exists in cwd", () => {

--- a/packages/server-build/__tests__/integration.test.ts
+++ b/packages/server-build/__tests__/integration.test.ts
@@ -55,9 +55,13 @@ describe("@paretools/build integration", () => {
 
   describe("tsc", () => {
     it("returns structured TypeScript diagnostics", async () => {
-      const repoRoot = resolve(__dirname, "../../..");
+      // Use this package's own dir so the binary-resolution check (#842) finds
+      // tsc in packages/server-build/node_modules/.bin/. pnpm doesn't hoist
+      // tsc to the repo-root node_modules in this workspace, so calling with
+      // path=repoRoot fails the assertBinaryAvailable check.
+      const packageRoot = resolve(__dirname, "..");
       const result = await client.callTool(
-        { name: "tsc", arguments: { path: repoRoot, noEmit: true } },
+        { name: "tsc", arguments: { path: packageRoot, noEmit: true } },
         undefined,
         CALL_TIMEOUT,
       );

--- a/packages/server-build/src/lib/build-runner.ts
+++ b/packages/server-build/src/lib/build-runner.ts
@@ -1,6 +1,46 @@
+import { existsSync } from "node:fs";
+import { dirname, join, parse } from "node:path";
 import { run, type RunResult } from "@paretools/shared";
 
+/**
+ * Walks up from `cwd` looking for `node_modules/.bin/<binary>` (or `<binary>.cmd`
+ * on Windows). Returns the resolved path, or `undefined` if not found before
+ * hitting the filesystem root.
+ */
+function findLocalBinary(cwd: string, binary: string): string | undefined {
+  const isWin = process.platform === "win32";
+  const candidates = isWin ? [`${binary}.cmd`, `${binary}.exe`, binary] : [binary];
+  let current = cwd;
+  const root = parse(current).root;
+  while (true) {
+    for (const name of candidates) {
+      const candidate = join(current, "node_modules", ".bin", name);
+      if (existsSync(candidate)) return candidate;
+    }
+    if (current === root) return undefined;
+    const parent = dirname(current);
+    if (parent === current) return undefined;
+    current = parent;
+  }
+}
+
+/**
+ * Asserts that a binary is available at `<cwd>/node_modules/.bin/<binary>` (or
+ * any ancestor — supports workspaces). Throws a typed error pointing at the
+ * missing-deps case so consumers don't get a contradictory empty diagnostics
+ * result. See #842.
+ */
+export function assertBinaryAvailable(cwd: string, binary: string): void {
+  if (findLocalBinary(cwd, binary)) return;
+  const expected = join(cwd, "node_modules", ".bin", binary);
+  throw new Error(
+    `${binary} binary not found at ${expected} — try running "pnpm install" (or "npm install" / "yarn install") in ${cwd}.`,
+  );
+}
+
 export async function tsc(args: string[], cwd?: string): Promise<RunResult> {
+  const workdir = cwd ?? process.cwd();
+  assertBinaryAvailable(workdir, "tsc");
   return run("npx", ["tsc", ...args], { cwd, timeout: 180_000 });
 }
 

--- a/packages/server-npm/__tests__/install-verification.test.ts
+++ b/packages/server-npm/__tests__/install-verification.test.ts
@@ -1,0 +1,92 @@
+import { describe, it, expect, beforeEach, afterEach } from "vitest";
+import { mkdtempSync, rmSync, mkdirSync } from "node:fs";
+import { tmpdir } from "node:os";
+import { join } from "node:path";
+import { assertInstallActuallyHappened } from "../src/tools/install.js";
+
+describe("assertInstallActuallyHappened", () => {
+  let tmp: string;
+
+  beforeEach(() => {
+    tmp = mkdtempSync(join(tmpdir(), "pare-npm-installverify-"));
+  });
+
+  afterEach(() => {
+    rmSync(tmp, { recursive: true, force: true });
+  });
+
+  it("throws when node_modules/ does not exist after install", () => {
+    expect(() =>
+      assertInstallActuallyHappened({
+        cwd: tmp,
+        pm: "pnpm",
+        subcommand: "install",
+        exitCode: 0,
+        stderr: "Already up to date\nDone in 1.0s",
+      }),
+    ).toThrowError(/no node_modules\/ was created/);
+  });
+
+  it("includes the pm, cwd, and exitCode in the diagnostic message", () => {
+    let caught: Error | undefined;
+    try {
+      assertInstallActuallyHappened({
+        cwd: tmp,
+        pm: "pnpm",
+        subcommand: "install",
+        exitCode: 0,
+        stderr: "Lockfile is up to date, resolution step is skipped",
+      });
+    } catch (err) {
+      caught = err as Error;
+    }
+    expect(caught).toBeDefined();
+    expect(caught!.message).toContain("pm=pnpm");
+    expect(caught!.message).toContain(`cwd=${tmp}`);
+    expect(caught!.message).toContain("exitCode=0");
+    expect(caught!.message).toContain("stderr (last 5 lines):");
+    expect(caught!.message).toContain("Lockfile is up to date");
+  });
+
+  it("does NOT throw when node_modules/ exists after install", () => {
+    mkdirSync(join(tmp, "node_modules"));
+    expect(() =>
+      assertInstallActuallyHappened({
+        cwd: tmp,
+        pm: "pnpm",
+        subcommand: "install",
+        exitCode: 0,
+        stderr: "",
+      }),
+    ).not.toThrow();
+  });
+
+  it("works for npm too", () => {
+    expect(() =>
+      assertInstallActuallyHappened({
+        cwd: tmp,
+        pm: "npm",
+        subcommand: "ci",
+        exitCode: 0,
+        stderr: "",
+      }),
+    ).toThrowError(/npm ci reported success but no node_modules\//);
+  });
+
+  it("omits the stderr block when stderr is empty", () => {
+    let caught: Error | undefined;
+    try {
+      assertInstallActuallyHappened({
+        cwd: tmp,
+        pm: "yarn",
+        subcommand: "install",
+        exitCode: 0,
+        stderr: "",
+      });
+    } catch (err) {
+      caught = err as Error;
+    }
+    expect(caught).toBeDefined();
+    expect(caught!.message).not.toContain("stderr (last 5 lines):");
+  });
+});

--- a/packages/server-npm/src/tools/install.ts
+++ b/packages/server-npm/src/tools/install.ts
@@ -1,6 +1,8 @@
 import { z } from "zod";
 import { createHash } from "node:crypto";
+import { existsSync } from "node:fs";
 import { readFile } from "node:fs/promises";
+import { join } from "node:path";
 import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
 import {
   dualOutput,
@@ -160,10 +162,52 @@ export function registerInstallTool(server: McpServer) {
 
       const output = result.stdout + "\n" + result.stderr;
       const install = parseInstallOutput(output);
+
+      // Sanity check: install (other than dry-run / global) MUST leave a
+      // node_modules/ on disk. Without this guard, a CLI that exits 0 with
+      // a benign "already up to date" message gets reported as success even
+      // though nothing was installed. See #842.
+      if (!dryRun && !isGlobal) {
+        assertInstallActuallyHappened({
+          cwd,
+          pm,
+          subcommand,
+          exitCode: result.exitCode,
+          stderr: result.stderr,
+        });
+      }
+
       return dualOutput({ ...install, packageManager: pm, lockfileChanged }, (d) =>
         formatInstall(d, duration),
       );
     },
+  );
+}
+
+/**
+ * Throws when an install command reports success but no `node_modules/` was
+ * created in `cwd`. This catches the dominant silent-success failure mode
+ * described in #842, where pnpm exits 0 without actually installing.
+ *
+ * Exported so unit tests can exercise the verification without needing to
+ * shell out to a real package manager.
+ */
+export function assertInstallActuallyHappened(args: {
+  cwd: string;
+  pm: "npm" | "pnpm" | "yarn";
+  subcommand: string;
+  exitCode: number;
+  stderr: string;
+}): void {
+  const nodeModulesPath = join(args.cwd, "node_modules");
+  if (existsSync(nodeModulesPath)) return;
+
+  const stderrTail = args.stderr.trim().split("\n").slice(-5).join("\n");
+  throw new Error(
+    `${args.pm} ${args.subcommand} reported success but no node_modules/ was created at ${nodeModulesPath}. ` +
+      `This usually means the package manager exited 0 without actually installing (e.g. a misleading "already up to date" message, or a workspace filter that matched nothing).\n` +
+      `pm=${args.pm} cwd=${args.cwd} exitCode=${args.exitCode}` +
+      (stderrTail ? `\nstderr (last 5 lines):\n${stderrTail}` : ""),
   );
 }
 

--- a/packages/server-test/__tests__/binary-check.test.ts
+++ b/packages/server-test/__tests__/binary-check.test.ts
@@ -1,0 +1,57 @@
+import { describe, it, expect, beforeEach, afterEach } from "vitest";
+import { mkdtempSync, rmSync, mkdirSync, writeFileSync, chmodSync } from "node:fs";
+import { tmpdir } from "node:os";
+import { join } from "node:path";
+import { assertNodeFrameworkAvailable } from "../src/lib/binary-check.js";
+
+describe("assertNodeFrameworkAvailable", () => {
+  let tmp: string;
+
+  beforeEach(() => {
+    tmp = mkdtempSync(join(tmpdir(), "pare-test-binarycheck-"));
+  });
+
+  afterEach(() => {
+    rmSync(tmp, { recursive: true, force: true });
+  });
+
+  it("throws a typed error when the framework binary is missing (vitest)", () => {
+    expect(() => assertNodeFrameworkAvailable(tmp, "vitest")).toThrowError(
+      /vitest binary not found at .*node_modules.*\.bin.*vitest.*— try running "pnpm install"/,
+    );
+  });
+
+  it("throws for jest when missing", () => {
+    expect(() => assertNodeFrameworkAvailable(tmp, "jest")).toThrowError(/jest binary not found/);
+  });
+
+  it("throws for mocha when missing", () => {
+    expect(() => assertNodeFrameworkAvailable(tmp, "mocha")).toThrowError(/mocha binary not found/);
+  });
+
+  it("never throws for pytest (lives outside node_modules)", () => {
+    expect(() => assertNodeFrameworkAvailable(tmp, "pytest")).not.toThrow();
+  });
+
+  it("succeeds when the binary exists in cwd's node_modules", () => {
+    const binDir = join(tmp, "node_modules", ".bin");
+    mkdirSync(binDir, { recursive: true });
+    const binPath = join(binDir, "vitest");
+    writeFileSync(binPath, "#!/usr/bin/env node\n");
+    chmodSync(binPath, 0o755);
+    expect(() => assertNodeFrameworkAvailable(tmp, "vitest")).not.toThrow();
+  });
+
+  it("succeeds when the binary is hoisted to an ancestor (workspace setup)", () => {
+    const binDir = join(tmp, "node_modules", ".bin");
+    mkdirSync(binDir, { recursive: true });
+    const binPath = join(binDir, "vitest");
+    writeFileSync(binPath, "#!/usr/bin/env node\n");
+    chmodSync(binPath, 0o755);
+
+    const nested = join(tmp, "packages", "child");
+    mkdirSync(nested, { recursive: true });
+
+    expect(() => assertNodeFrameworkAvailable(nested, "vitest")).not.toThrow();
+  });
+});

--- a/packages/server-test/src/lib/binary-check.ts
+++ b/packages/server-test/src/lib/binary-check.ts
@@ -1,0 +1,43 @@
+import { existsSync } from "node:fs";
+import { dirname, join, parse } from "node:path";
+import type { Framework } from "./detect.js";
+
+/**
+ * Walks up from `cwd` looking for `node_modules/.bin/<binary>` (or `<binary>.cmd`
+ * / `<binary>.exe` on Windows). Returns the first match, or `undefined` when no
+ * ancestor of `cwd` provides the binary.
+ */
+function findLocalBinary(cwd: string, binary: string): string | undefined {
+  const isWin = process.platform === "win32";
+  const candidates = isWin ? [`${binary}.cmd`, `${binary}.exe`, binary] : [binary];
+  let current = cwd;
+  const root = parse(current).root;
+  while (true) {
+    for (const name of candidates) {
+      const candidate = join(current, "node_modules", ".bin", name);
+      if (existsSync(candidate)) return candidate;
+    }
+    if (current === root) return undefined;
+    const parent = dirname(current);
+    if (parent === current) return undefined;
+    current = parent;
+  }
+}
+
+/**
+ * Asserts that the JS-side test framework binary (jest, vitest, mocha) is
+ * resolvable from `cwd`. Throws a typed, actionable error when it isn't —
+ * this is the dominant failure mode when a fresh worktree has no
+ * `node_modules/`. See #842.
+ *
+ * pytest is intentionally skipped because it's invoked via `python -m pytest`
+ * and lives outside the npm `node_modules/.bin` resolution path.
+ */
+export function assertNodeFrameworkAvailable(cwd: string, framework: Framework): void {
+  if (framework === "pytest") return;
+  if (findLocalBinary(cwd, framework)) return;
+  const expected = join(cwd, "node_modules", ".bin", framework);
+  throw new Error(
+    `${framework} binary not found at ${expected} — try running "pnpm install" (or "npm install" / "yarn install") in ${cwd}.`,
+  );
+}

--- a/packages/server-test/src/tools/run.ts
+++ b/packages/server-test/src/tools/run.ts
@@ -15,6 +15,7 @@ import {
   coerceJsonArray,
 } from "@paretools/shared";
 import { detectFramework, type Framework } from "../lib/detect.js";
+import { assertNodeFrameworkAvailable } from "../lib/binary-check.js";
 import { parsePytestOutput } from "../lib/parsers/pytest.js";
 import { parseJestJson } from "../lib/parsers/jest.js";
 import { parseVitestJson } from "../lib/parsers/vitest.js";
@@ -391,6 +392,7 @@ export function registerRunTool(server: McpServer) {
 
       const cwd = path || process.cwd();
       const detected = framework || (await detectFramework(cwd));
+      assertNodeFrameworkAvailable(cwd, detected);
       const extraArgs = buildRunExtraArgs(detected, {
         filter,
         shard,

--- a/tests/smoke/mocked/npm-tools.smoke.test.ts
+++ b/tests/smoke/mocked/npm-tools.smoke.test.ts
@@ -42,6 +42,16 @@ vi.mock("node:fs/promises", () => ({
   access: vi.fn().mockRejectedValue(new Error("ENOENT")),
 }));
 
+// Mock node:fs's existsSync so the post-install verification (#842) treats the
+// fake mock paths (e.g. /tmp/project) as if node_modules/ exists. The unit
+// tests in packages/server-npm/__tests__/install-verification.test.ts cover
+// the real existsSync-based assertion behavior. server-npm only imports
+// existsSync from node:fs (verified by grep), so the minimal mock below
+// is safe — no other fs functions need to pass through.
+vi.mock("node:fs", () => ({
+  existsSync: vi.fn().mockReturnValue(true),
+}));
+
 // Mock @paretools/shared run for nvm tool
 vi.mock("@paretools/shared", async () => {
   const actual = await vi.importActual<typeof import("@paretools/shared")>("@paretools/shared");
@@ -77,6 +87,7 @@ import { registerOutdatedTool } from "../../../packages/server-npm/src/tools/out
 import { registerRunTool } from "../../../packages/server-npm/src/tools/run.js";
 import { registerSearchTool } from "../../../packages/server-npm/src/tools/search.js";
 import { registerTestTool } from "../../../packages/server-npm/src/tools/test.js";
+import { existsSync } from "node:fs";
 
 // ── Types & Helpers ────────────────────────────────────────────────────────
 
@@ -583,6 +594,10 @@ describe("Smoke: npm.install", () => {
     vi.resetAllMocks();
     vi.mocked(detectPackageManager).mockResolvedValue("npm");
     vi.mocked(readFile).mockRejectedValue(new Error("ENOENT"));
+    // The post-install verification (#842) calls existsSync(<cwd>/node_modules).
+    // Smoke tests use mock paths like /tmp/project that don't exist on disk —
+    // re-set after vi.resetAllMocks() above wipes the file-level mock factory.
+    vi.mocked(existsSync).mockReturnValue(true);
     const server = new FakeServer();
     registerInstallTool(server as never);
     handler = server.tools.get("install")!.handler;

--- a/tests/smoke/mocked/test-tools.smoke.test.ts
+++ b/tests/smoke/mocked/test-tools.smoke.test.ts
@@ -44,6 +44,18 @@ vi.mock("node:fs/promises", async (importOriginal) => {
   };
 });
 
+// Mock node:fs's existsSync so the framework-binary check (#842) treats the
+// fake mock paths (e.g. /project) as if node_modules/.bin/<framework> exists.
+// The unit tests in packages/server-test/__tests__/binary-check.test.ts cover
+// the real existsSync-based assertion behavior.
+vi.mock("node:fs", async (importOriginal) => {
+  const actual = (await importOriginal()) as Record<string, unknown>;
+  return {
+    ...actual,
+    existsSync: vi.fn().mockReturnValue(true),
+  };
+});
+
 import { run } from "../../../packages/shared/dist/runner.js";
 import { detectFramework } from "../../../packages/server-test/src/lib/detect.js";
 import { readFile } from "node:fs/promises";

--- a/tests/smoke/suite/test-tools.recorded.test.ts
+++ b/tests/smoke/suite/test-tools.recorded.test.ts
@@ -36,6 +36,17 @@ vi.mock("node:fs/promises", async (importOriginal) => {
   };
 });
 
+// Mock node:fs's existsSync so the framework-binary check (#842) treats the
+// fake mock paths as if node_modules/.bin/<framework> exists. Preserves
+// readFileSync (used to load fixtures) by spreading the actual module first.
+vi.mock("node:fs", async (importOriginal) => {
+  const actual = (await importOriginal()) as Record<string, unknown>;
+  return {
+    ...actual,
+    existsSync: vi.fn().mockReturnValue(true),
+  };
+});
+
 import { run } from "../../../packages/shared/dist/runner.js";
 import { detectFramework } from "../../../packages/server-test/src/lib/detect.js";
 import { readFile, unlink, mkdir, rm } from "node:fs/promises";

